google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1alpha2/JsonStreamWriterTest.java

```diff
@@ -0,0 +1,611 @@
+    testBigQueryWrite.addResponse(
+        Storage.AppendRowsResponse.newBuilder()
+            .setError(com.google.rpc.Status.newBuilder().setCode(11).build())
+            .build());
+    JSONObject foo = new JSONObject();
+    foo.put("foo", "allen");
+    JSONArray jsonArr = new JSONArray();
+    jsonArr.put(foo);
+    ApiFuture<AppendRowsResponse> appendFuture =
+        writer.append(jsonArr, -1, /* allowUnknownFields */ false);
+    try {
+      appendFuture.get();
+    } catch (Throwable t) {
+      assertEquals(t.getCause().getMessage(), "OUT_OF_RANGE: ");
+    }
+    writer.close();
+  }
+
+  @Test
+  public void testAppendOutOfRangeAndUpdateSchema() throws Exception {
+    JsonStreamWriter writer = getTestJsonStreamWriterBuilder(TEST_STREAM, TABLE_SCHEMA).build();
+    testBigQueryWrite.addResponse(
+        Storage.AppendRowsResponse.newBuilder()
+            .setError(com.google.rpc.Status.newBuilder().setCode(11).build())
+            .setUpdatedSchema(UPDATED_TABLE_SCHEMA)
+            .build());
+    testBigQueryWrite.addResponse(Storage.AppendRowsResponse.newBuilder().setOffset(0).build());
+
+    JSONObject foo = new JSONObject();
+    foo.put("foo", "allen");
+    JSONArray jsonArr = new JSONArray();
+    jsonArr.put(foo);
+    ApiFuture<AppendRowsResponse> appendFuture =
+        writer.append(jsonArr, -1, /* allowUnknownFields */ false);
+    try {
+      appendFuture.get();
+    } catch (Throwable t) {
+      assertEquals(t.getCause().getMessage(), "OUT_OF_RANGE: ");
+      int millis = 0;
+      while (millis <= 10000) {
+        if (writer.getDescriptor().getFields().size() == 2) {
+          break;
+        }
+        Thread.sleep(100);
+        millis += 100;
+      }
+      assertTrue(writer.getDescriptor().getFields().size() == 2);
+    }
+
+    JSONObject updatedFoo = new JSONObject();
+    updatedFoo.put("foo", "allen");
+    updatedFoo.put("bar", "allen2");
+    JSONArray updatedJsonArr = new JSONArray();
+    updatedJsonArr.put(updatedFoo);
+
+    ApiFuture<AppendRowsResponse> appendFuture2 =
+        writer.append(updatedJsonArr, -1, /* allowUnknownFields */ false);
+
+    assertEquals(0L, appendFuture2.get().getOffset());
+    assertEquals(
+        1,
+        testBigQueryWrite
+            .getAppendRequests()
+            .get(1)
+            .getProtoRows()
+            .getRows()
+            .getSerializedRowsCount());
+    assertEquals(
+        testBigQueryWrite.getAppendRequests().get(1).getProtoRows().getRows().getSerializedRows(0),
+        UpdatedFooType.newBuilder().setFoo("allen").setBar("allen2").build().toByteString());
+
+    // Check if writer schemas were added in for both connections.
+    assertTrue(testBigQueryWrite.getAppendRequests().get(0).getProtoRows().hasWriterSchema());
+    assertTrue(testBigQueryWrite.getAppendRequests().get(1).getProtoRows().hasWriterSchema());
+
+    writer.close();
+  }
+
+  @Test
+  public void testMultiThreadAppend() throws Exception {
```

Comment:
How about drop in some schema update in between these parallel calls.

Reply:
Will do!